official/recommendation/ncf_keras_main.py

```diff
@@ -0,0 +1,241 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""NCF framework to train and evaluate the NeuMF model.
+
+The NeuMF model assembles both MF and MLP models under the NCF framework. Check
+`neumf_model.py` for more details about the models.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import contextlib
+import heapq
+import json
+import logging
+import math
+import multiprocessing
+import os
+import signal
+import typing
+
+# pylint: disable=g-bad-import-order
+import numpy as np
+from absl import app as absl_app
+from absl import flags
+import tensorflow as tf
+# pylint: enable=g-bad-import-order
+
+from tensorflow.contrib.compiler import xla
+from official.datasets import movielens
+from official.recommendation import constants as rconst
+from official.recommendation import data_pipeline
+from official.recommendation import data_preprocessing
+from official.recommendation import ncf_common
+from official.recommendation import neumf_model
+from official.utils.flags import core as flags_core
+from official.utils.logs import hooks_helper
+from official.utils.logs import logger
+from official.utils.logs import mlperf_helper
+from official.utils.misc import distribution_utils
+from official.utils.misc import model_helpers
+
+
+FLAGS = flags.FLAGS
+
+
+def main(_):
+  with logger.benchmark_context(FLAGS), \
+      mlperf_helper.LOGGER(FLAGS.output_ml_perf_compliance_logging):
+    mlperf_helper.set_ncf_root(os.path.split(os.path.abspath(__file__))[0])
+    if FLAGS.tpu:
+      raise ValueError("NCF in Keras does not support TPU for now")
+    run_ncf(FLAGS)
+
+
+def run_ncf(_):
+  """Run NCF training and eval with Keras."""
+  params = ncf_common.parse_flags(FLAGS)
+
+  distribution = ncf_common.get_distribution_strategy(params)
+
+  num_users, num_items, num_train_steps, num_eval_steps, producer = (
+      ncf_common.get_inputs(params))
+
+  params["num_users"], params["num_items"] = num_users, num_items
+  producer.start()
+  model_helpers.apply_clean(flags.FLAGS)
+
+  # TODO(shiningsun): Both MirroredStrategy and OneDeviceStrategy error out.
+  # Find out why and change the distribute to distribution
+  with distribution_utils.MaybeDistributionScope(None):
+    keras_model = _get_keras_model(params)
```

Comment:
Would be good to use explicit keyword args instead of "params" to improve readability